### PR TITLE
Request based PoV distribution

### LINKED ISSUE #2590: Request based PoV distribution
Right now gossip via notification protocol.

PoVs are large so this is bad for two reasons:

- Gossip is not ideal for transmission of huge payloads
- Notification protocols are also not optimized for sending of huge amounts of data.

### PR DESCRIPTION
**Pull request by eskimor**
Fixes #2590

### CODE REVIEW

File: node/network/availability-distribution/src/lib.rs [hunk 1]
```diff
@@ -107,14 +120,14 @@ impl AvailabilityDistributionSubsystem {
...
 				FromOverseer::Signal(OverseerSignal::ActiveLeaves(update)) => {
-					// Update the relay chain heads we are fetching our pieces for:
-					if let Some(e) = requester
-						.get_mut()
-						.update_fetching_heads(&mut ctx, update)
-						.await?
-					{
-						tracing::debug!(target: LOG_TARGET, "Error processing ActiveLeavesUpdate: {:?}", e);
-					}
+                    log_error(
```
Comment (rphmeier) [inline node/network/availability-distribution/src/lib.rs:123]: spaces i believe

File: node/network/bridge/src/network.rs [hunk 2]
```diff
@@ -218,14 +218,14 @@ impl Network for Arc<NetworkService<Block, Hash>> {
...
 				authority_discovery
-				.get_addresses_by_authority_id(authority)
-				.await
-				.and_then(|addrs| {
-					addrs
-						.into_iter()
-						.find_map(|addr| peer_id_from_multiaddr(&addr))
-				}),
-			};
+                    .get_addresses_by_authority_id(authority)
```
Comment (rphmeier) [inline node/network/bridge/src/network.rs:221]: and here

File: node/subsystem/src/messages.rs [hunk 3]
```diff
@@ -260,10 +260,27 @@ impl NetworkBridgeMessage {
...
 pub enum AvailabilityDistributionMessage {
 	/// Incoming network request for an availability chunk.
-	AvailabilityFetchingRequest(IncomingRequest<req_res_v1::AvailabilityFetchingRequest>)
+	AvailabilityFetchingRequest(IncomingRequest<req_res_v1::AvailabilityFetchingRequest>),
+	/// Incoming network request for a seconded PoV.
+	PoVFetchingRequest(IncomingRequest<req_res_v1::PoVFetchingRequest>),
+	/// Instruct availability distribution to fetch a remote PoV.
+	///
+	/// NOTE: The result of this fetch is not yet locally validated and could be bogus.
+	FetchPoV{
```
Comment (rphmeier) [inline node/subsystem/src/messages.rs:272]:
```suggestion
	FetchPoV {
```

File: node/network/availability-distribution/src/runtime.rs [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+			Ok(Some(info))
+		} else {
+			Ok(None)
+		}
+	}
+
+	/// Get our `ValidatorIndex`.
+	///
+	/// Returns: None if we are not a validator.
+	async fn get_our_index(&self, validators: &Vec<ValidatorId>) -> Option<ValidatorIndex> {
```
Comment (rphmeier) [inline node/network/availability-distribution/src/runtime.rs:189]:
```suggestion
	async fn get_our_index(&self, validators: &[ValidatorId]) -> Option<ValidatorIndex> {
```

File: node/network/availability-distribution/src/pov_requester/mod.rs [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+			},
+		);
+		let full_req = Requests::PoVFetching(req);
+
+		ctx.send_message(
+			AllMessages::NetworkBridge(
+				NetworkBridgeMessage::SendRequests(
+					vec![full_req],
+					// We are supposed to be connected to validators of our group:
+					IfDisconnected::ImmediateError
```
Comment (rphmeier) [inline node/network/availability-distribution/src/pov_requester/mod.rs:116]: We are supposed to be, but it still seems better not to have immediate errors and continue on trying to connect

File: node/network/availability-distribution/src/lib.rs [hunk 6]
```diff
@@ -123,8 +136,33 @@ impl AvailabilityDistributionSubsystem {
...
-					answer_request_log(&mut ctx, req, &self.metrics).await
+					answer_chunk_request_log(&mut ctx, req, &self.metrics).await
+				}
+				FromOverseer::Communication {
+					msg: AvailabilityDistributionMessage::PoVFetchingRequest(req),
+				} => {
+					answer_pov_request_log(&mut ctx, req, &self.metrics).await
 				}
+				FromOverseer::Communication {
+					msg: AvailabilityDistributionMessage::FetchPoV{
```
Comment (rphmeier) [inline node/network/availability-distribution/src/lib.rs:147]:
```suggestion
					msg: AvailabilityDistributionMessage::FetchPoV {
```

File: node/network/availability-distribution/src/lib.rs [hunk 7]
```diff
@@ -123,8 +136,33 @@ impl AvailabilityDistributionSubsystem {
...
+                        relay_parent,
+                        from_validator,
+                        candidate_hash,
+                        tx,
+                    },
+				} => {
+                    log_error(
+                        pov_requester.fetch_pov(
```
Comment (rphmeier) [inline node/network/availability-distribution/src/lib.rs:155]: The previous API I think was better for backing - request the PoV from _someone_. Now backing has to worry about from whom. And if it fails, retry, for another validator.

File: node/core/backing/src/lib.rs [hunk 8]
```diff
@@ -336,18 +355,51 @@ async fn make_pov_available(
...
+		)
+	))).await?;
+
+	let response = pending_response.await.map_err(Error::FetchPoV)?;
+	let pov = match response {
+		PoVFetchingResponse::PoV(compressed) => {
+			compressed.decompress().map_err(|_| Error::FetchPoVInvalidResponse)?
+		}
+		PoVFetchingResponse::NoSuchPoV => {
+			return Err(Error::FetchPoVNoSuchPoV)
```
Comment (rphmeier) [inline node/core/backing/src/lib.rs:399]: It's really 'no such PoV from this validator but someone else might have it'

File: node/core/backing/src/lib.rs [hunk 9]
```diff
@@ -336,18 +355,51 @@ async fn make_pov_available(
...
+	let full_req = Requests::PoVFetching(req);
+
+	tx_from.send(FromJobCommand::SendMessage(AllMessages::NetworkBridge(
+		NetworkBridgeMessage::SendRequests(
+			vec![full_req],
+			IfDisconnected::ImmediateError
+		)
+	))).await?;
+
+	let response = pending_response.await.map_err(Error::FetchPoV)?;
```
Comment (rphmeier) [inline node/core/backing/src/lib.rs:393]: This fails if the validator is not already connected also. We'd need retrying logic in backing

**Review by rphmeier — COMMENTED** [review]
I think it would be better to keep the same external API of "get the PoV from someone in the group" as it keeps backing simple. Retrying & querying different peers can be handled within the subsystem.

Also, I don't think availability distribution is really the best home for this logic as it has little to do with availability.

File: node/core/backing/src/lib.rs (hunk 8 above)
Comment (eskimor) [reply node/core/backing/src/lib.rs:399]: Uh forget the code in backing - that's old. Fixing it today.

**Comment by eskimor** [comment]
> Also, I don't think availability distribution is really the best home for this logic as it has little to do with availability and everything to do with backing

Yeah, we would probably need a different name if we keep it that way. Pierre's blog post inspired me to an architecture, where we stay in "standard" Rust more, with easy sharing of data, via references and such and still get all the benefits of asynchronous programming and the efficiency/concurrency it brings. Basically simplifying things a lot/making it easier to reason about code and also improving performance. I wanted to try this out in the small here and give a knowledge share presentation if it worked out, to me it looks very promising, but only code does not lie.

File: node/network/availability-distribution/src/lib.rs (hunk 7 above)
Comment (eskimor) [reply node/network/availability-distribution/src/lib.rs]: True, we could make pov distribution also listen to statements, that would be the cleanest approach to avoiding that, I think. Although, I have already implemented retry logic in backing now.

File: node/network/availability-distribution/src/pov_requester/mod.rs [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+	parent: Hash,
+	session: SessionIndex
+)
+	-> super::Result<Option<mpsc::Receiver<(AuthorityDiscoveryId, PeerId)>>>
+where
+	Context: SubsystemContext,
+{
+	if let Some(validator_ids) = determine_relevant_validators(ctx, runtime, parent, session).await? {
+		// We don't actually care about `PeerId`s, just keeping receiver so we stay connected:
+		let (tx, rx) = mpsc::channel(0);
```
Comment (rphmeier) [inline node/network/availability-distribution/src/pov_requester/mod.rs:190]: might this block the validator-discovery worker somehow?

File: node/network/availability-distribution/src/requester/mod.rs [hunk 11]
```diff
@@ -238,7 +224,7 @@ impl Stream for Requester {
 async fn query_occupied_cores<Context>(
 	ctx: &mut Context,
 	relay_parent: Hash,
-) -> Result<Vec<OccupiedCore>, NonFatalError>
+) -> Result<Vec<OccupiedCore>, Error>
```
Comment (rphmeier) [inline node/network/availability-distribution/src/requester/mod.rs:227]: why this change from `NonFatalError`?

File: node/network/availability-distribution/src/requester/mod.rs [hunk 12]
```diff
@@ -112,35 +112,25 @@ impl Requester {
...
-			tracing::trace!(
-				target: LOG_TARGET,
-				occupied_cores = ?cores,
-				"Query occupied core"
-			);
-			if let Some(err) = self.add_cores(ctx, leaf, cores).await? {
-				return Ok(Some(err));
-			}
+			let cores = query_occupied_cores(ctx, leaf).await?;
+			self.add_cores(ctx, leaf, cores).await?;
```
Comment (rphmeier) [inline node/network/availability-distribution/src/requester/mod.rs:131]: Some tracing was lost

File: node/network/availability-distribution/src/responder.rs [hunk 13]
```diff
@@ -28,44 +28,108 @@ use polkadot_subsystem::{
...
 
-/// Variant of `answer_request` that does Prometheus metric and logging on errors.
+/// Variant of `answer_pov_request` that does Prometheus metric and logging on errors.
+///
+/// Any errors of `answer_pov_request` will simply be logged.
+pub async fn answer_pov_request_log<Context>(
+	ctx: &mut Context,
+	req: IncomingRequest<v1::PoVFetchingRequest>,
+	metrics: &Metrics,
+) -> ()
```
Comment (rphmeier) [inline node/network/availability-distribution/src/responder.rs:38]: explicit `()` return type not necessary

File: node/network/availability-distribution/src/pov_requester/mod.rs (hunk 10 above)
Comment (eskimor) [reply node/network/availability-distribution/src/pov_requester/mod.rs:190]: No - I checked, it is smart enough to do a try_send.

**Review by rphmeier — COMMENTED** [review]
Guide updates, now that PoV distribution is removed?

And there should be a note that it can only successfully gather the data from the store because it is placed there after attesting/seconding.

File: node/network/availability-distribution/src/requester/mod.rs (hunk 11 above)
Comment (eskimor) [reply node/network/availability-distribution/src/requester/mod.rs:227]: It was inconsistent and did not feel useful at all. (Like we still did not fail on lots of fatal `Error`s)  The distinction felt pretty arbitrary, also what I got, we basically never want to fail, which makes that distinction even less useful. If there are circumstances we really want to crash, we can panic, which should be pretty much the same as propagating `Err` to the overseer.

Comment (ordian) [reply node/network/availability-distribution/src/requester/mod.rs:227]: I guess the distinction was about whether to log an error or not. We probably don't want to log a runtime api error for example.

Comment (eskimor) [reply node/network/availability-distribution/src/requester/mod.rs:232]: I think we should do a `debug_assert` on logic errors or if really bad an `assert`, this should get us the same thing, but with less complexity.

Comment (ordian) [reply node/network/availability-distribution/src/requester/mod.rs:232]: I see 3 classes of errors:
* NonFatal - log and continue - atm it's only RuntimeApiError
* Logic - debug_assert/assert and warn
* ChannelClosed - a subsystem is down - we're shutting down - early exit

Comment (ordian) [reply node/network/availability-distribution/src/requester/mod.rs:232]: panicking is usually bad because we may trigger double panic which would lead to a segfault which may be hard to debug and some subsystems might prefer to have a clean shutdown to e.g. flush a db transaction

Comment (eskimor) [reply node/network/availability-distribution/src/requester/mod.rs:232]: Hmm. Then I think we could do a trait `IsFatal`with a function `isFatal` and have a generic function that just logs non fatal errors and will propagate fatal ones, that would be the least invasive method I guess. 

Like, changing what is fatal and what is not can be done in a single place and we could have a default implementation for `isFatal` which just always returns `false`.

Comment (ordian) [reply node/network/availability-distribution/src/requester/mod.rs:232]: There were some attempts to do this systematically (starting with #1888), but we never got around enforcing this on a compile time level as there were always more important stuff to do :P

I agree though that'd be really nice to have in one place.

**Review by rphmeier — APPROVED** [review]

**Comment by zqhxuyuan** [comment]
@eskimor any update on implements guide? the doc still has pov-distribution here: https://w3f.github.io/parachain-implementers-guide/node/subsystems-and-jobs.html

**Comment by eskimor** [comment]
I'll revisit the guide, thanks!
